proofs/proof.ml

```diff
@@ -365,34 +358,46 @@ let run_tactic env tac pr =
   let open Proofview.Notations in
   let undef sigma l = List.filter (fun g -> Evd.is_undefined sigma g) l in
   let tac =
+    Proofview.tclEVARMAP >>= fun sigma ->
+    Proofview.Unsafe.tclEVARS (Evd.push_shelf sigma) >>= fun () ->
     tac >>= fun result ->
     Proofview.tclEVARMAP >>= fun sigma ->
     (* Already solved goals are not to be counted as shelved. Nor are
       they to be marked as unresolvable. *)
     let retrieved, sigma = Evd.pop_future_goals sigma in
     let retrieved = Evd.FutureGoals.filter (Evd.is_undefined sigma) retrieved in
-    let retrieved = List.rev_append retrieved.Evd.FutureGoals.shelf (List.rev retrieved.Evd.FutureGoals.comb) in
+    let retrieved = List.rev retrieved.Evd.FutureGoals.comb in
     let sigma = Proofview.Unsafe.mark_as_goals sigma retrieved in
+    let to_shelve, sigma = Evd.pop_shelf sigma in
     Proofview.Unsafe.tclEVARS sigma >>= fun () ->
-    Proofview.tclUNIT (result,retrieved)
+    Proofview.tclUNIT (result,retrieved,to_shelve)
   in
   let { name; poly; proofview } = pr in
   let proofview = Proofview.Unsafe.push_future_goals proofview in
-  let ((result,retrieved),proofview,(status,to_shelve),info_trace) =
+  let ((result,retrieved,to_shelve),proofview,status,info_trace) =
     Proofview.apply ~name ~poly env tac proofview
   in
   let sigma = Proofview.return proofview in
   let to_shelve = undef sigma to_shelve in
-  let shelf = (undef sigma pr.shelf)@retrieved@to_shelve in
   let proofview = Proofview.Unsafe.mark_as_unresolvables proofview to_shelve in
-  { pr with proofview ; shelf },(status,info_trace),result
+  let modify = function
+    | [] -> CErrors.anomaly Pp.(str"shelf stack should not be empty")
+    | hd :: q ->
+      (hd@retrieved@to_shelve) :: q
```

Comment:
I never understood this @retrieved here, why are the (unsolved) future goals comb considered as shelved here?

Reply:
Isn't this because the goals generated by tactics through the pretyper need to be shelved? Note that this PR does not change previous behavior for these "retrieved" goals.

Reply:
You are right, `tac` might have used the pretyper to typecheck a term and all the goal evars there should be shelved by default I suppose. Tactics such as `refine` manipulate the future goals so that actual subgoals do not appear in the future goals anymore. I'm now wondering how often tactics produce such future goals though (e.g. apply, rewrite). I think the main bug the new apply of unifall fixes is related: it tries to shelve a goal of apply that was not unified which is an error (for apply, not eapply), where the old apply doesn't realize that an evar under its responsibility is leaked to the shelf.

In any case, I think I understand everything a bit better now thanks to the refactorings :) Great job!